pkg/cmd/server/start/start_node.go

```diff
@@ -218,43 +214,86 @@ func (o NodeOptions) RunNode() error {
 		}
 	}
 	if len(validationResults.Errors) != 0 {
+		glog.V(4).Infof("Configuration is invalid: %#v", nodeConfig)
 		return kerrors.NewInvalid(configapi.Kind("NodeConfig"), o.ConfigFile, validationResults.Errors)
 	}
 
 	if err := ValidateRuntime(nodeConfig, o.NodeArgs.Components); err != nil {
 		return err
 	}
 
+	if o.IsWriteConfigOnly() {
+		return nil
+	}
+
 	if err := StartNode(*nodeConfig, o.NodeArgs.Components); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (o NodeOptions) CreateNodeConfig() error {
-	getSignerOptions := &admin.SignerCertOptions{
-		CertFile:   admin.DefaultCertFilename(o.NodeArgs.MasterCertDir, admin.CAFilePrefix),
-		KeyFile:    admin.DefaultKeyFilename(o.NodeArgs.MasterCertDir, admin.CAFilePrefix),
-		SerialFile: admin.DefaultSerialFilename(o.NodeArgs.MasterCertDir, admin.CAFilePrefix),
+// resolveNodeConfig creates a new configuration on disk by reading from the master, reads
+// the config file from disk if specified, or generates a new config from the incoming arguments.
+// After this call returns without an error, config files will exist on disk.
+func (o NodeOptions) resolveNodeConfig() (*configapi.NodeConfig, error) {
+	if o.NodeArgs.Bootstrap {
+		glog.V(2).Infof("Bootstrapping from master configuration")
+
+		hostnames, err := o.NodeArgs.GetServerCertHostnames()
+		if err != nil {
+			return nil, err
+		}
+		nodeConfigDir := o.NodeArgs.ConfigDir.Value()
+		if err := o.loadBootstrap(hostnames.List(), nodeConfigDir); err != nil {
+			return nil, err
+		}
+		if len(o.ConfigFile) == 0 {
+			o.ConfigFile = filepath.Join(o.NodeArgs.ConfigDir.Value(), "node-config.yaml")
+		}
+		return configapilatest.ReadAndResolveNodeConfig(o.ConfigFile)
+	}
+
```

Comment:
NodeOptions#Validate prevents IsRunFromConfig() and IsWriteConfigOnly() from both being true. Since you're reloading and validating the written config in the IsWriteConfigOnly case, I think all the following code can simplify to:

```
if o.IsRunFromConfig() {
	glog.V(2).Infof("Reading node configuration from %s", o.ConfigFile)
	return configapilatest.ReadAndResolveNodeConfig(o.ConfigFile)
}

glog.V(2).Infof("Generating node configuration")
configFile, err := o.createNodeConfig()
if err != nil {
  return nil, err
}
return configapilatest.ReadAndResolveNodeConfig(configFile)
```